ee/control/client_http.go

```diff
@@ -40,77 +63,150 @@ func NewControlHTTPClient(logger log.Logger, addr string, client *http.Client, o
-		)
-		return nil, err
-
-	case http.StatusNotModified:
-		// The control server sends back a 304 Not Modified status, without a body, which tells
-		// the client that the cached version of the response is still good to use
-		level.Debug(c.logger).Log(
-			"msg", "got HTTP 304 making control server request",
-		)
-		return nil, err
-	}
-
-	if response.StatusCode != http.StatusOK {
-		level.Error(c.logger).Log(
-			"msg", "got not-ok status code from control server",
-			"response_code", response.StatusCode,
-			"response_body", response.Body,
-		)
-		return nil, err
-	}
-
-	// response.Body will be closed before this function exits, get all the data now
-	body, err := io.ReadAll(response.Body)
+		return nil, fmt.Errorf("could not create config request: %w", err)
+	}
+	configReq.Header.Set(HeaderChallenge, string(challenge))
+	configReq.Header.Set("Content-Type", "application/json")
+	configReq.Header.Set("Accept", "application/json")
+
+	// Calculate first signature
+	localDbKeys := agent.LocalDbKeys()
+	if localDbKeys.Type() != "noop" {
+		key1, err := keyHeaderValue(localDbKeys)
+		if err != nil {
+			return nil, fmt.Errorf("could not get key header from local db keys: %w", err)
+		}
+		sig1, err := signatureHeaderValue(localDbKeys, challenge)
+		if err != nil {
+			return nil, fmt.Errorf("could not get signature header from local db keys: %w", err)
+		}
+		configReq.Header.Set(HeaderKey, key1)
+		configReq.Header.Set(HeaderSignature, sig1)
+	}
+
+	// Calculate second signature if available
+	hardwareKeys := agent.HardwareKeys()
+	if hardwareKeys.Type() != "noop" {
+		key2, err := keyHeaderValue(hardwareKeys)
+		if err != nil {
+			return nil, fmt.Errorf("could not get key header from hardware keys: %w", err)
+		}
+		sig2, err := signatureHeaderValue(hardwareKeys, challenge)
+		if err != nil {
+			return nil, fmt.Errorf("could not get signature header from hardware keys: %w", err)
+		}
+		configReq.Header.Set(HeaderKey2, key2)
+		configReq.Header.Set(HeaderSignature2, sig2)
+	}
+
+	configAndAuthKeyRaw, err := c.do(configReq)
 	if err != nil {
-		level.Debug(c.logger).Log(
-			"msg", "error reading response body from control server",
-			"err", err,
-		)
-		return nil, err
+		return nil, fmt.Errorf("could not make config request: %w", err)
+	}
+
+	var cfgResp configResponse
+	if err := json.Unmarshal(configAndAuthKeyRaw, &cfgResp); err != nil {
+		return nil, fmt.Errorf("could not unmarshal challenge response: %w", err)
 	}
 
-	reader := bytes.NewReader(body)
+	// Set the auth token for use when fetching objects by their hashes later
+	c.token = cfgResp.Token
+
+	reader := bytes.NewReader(cfgResp.Config)
 	return reader, nil
 }
 
-func (c *HTTPClient) do(verb, path string) (*http.Response, error) {
-	request, err := http.NewRequest(
-		verb,
-		c.url(path).String(),
-		nil,
-	)
+func (c *HTTPClient) GetSubsystemData(hash string) (io.Reader, error) {
+	if c.token == "" {
+		_, err := c.GetConfig()
```

Comment:
It may be better to just error out here instead of trying to handle an empty token. The caller needs to get hashes from the data returned by `GetConfig`, so if this function is being called without `GetConfig` called first, the hash is potentially stale or otherwise invalid.

Reply:
Oh yup that absolutely makes more sense, I'll pull this out!